### DISCUSSION TITLE
Best practices for sharing a schema between server and client

### DISCUSSION THREAD

**Discussion by kaysond** [MichalLytek/type-graphql#1199] [Q&A]
I'm developing an app that uses type-graphql + apollo server on the backend, and the apollo client + react for the frontend. Right now I have my repo split into 3 packages: client, server, and schema.

Server imports the schema, creates the resolver for it, and then does a `buildSchema()` for the `ApolloServer`.

On the client side, I import just the schema as types, and pass them as the generic type to `ApolloClient.useQuery<Type>()`.

This generally seems to work well. **Is there a different or better way to do things?**

One big problem I've noticed (though maybe its more specific to Apollo than sharing the schema), is that the type returned from `useQuery` is ignorant of the actual query itself. So if my type/schema is
```typescript
@ObjectType()
class Recipe {
  @Field(type => ID)
  id: string;

  @Field()
  title: string;

  @Field({ nullable: true })
  description?: string;

  @Field()
  creationDate: Date;

  @Field(type => [String])
  ingredients: string[];
}
```
and my query is
```graphql
query GetRecipeTitle {
  recipe {
    id
    title
  }
}
 ```
The returned `data` object from `useQuery()` only actually has the `id` and `title` properties, but because the schema type has `description` `creationDate` and `ingredients` as well, typescript things that those exist on `data` even though they don't. **Is there any way to address this?**

**Comment by MichalLytek** [comment]
TypeGraphQL produces standard GraphQL schema. You can you all the tools for GraphQL for frontend to have typesafe queries, like graphql code generator.

**Reply by kaysond** [reply]
Can you elaborate? Do you mean run `buildSchema` with `emitSchemaFile` then run something like graphql code generator with a script or at build time?

Right now, I have my type/schema in the `schema` package, and the `server` package imports that and creates the `resolver` for it which is then passed to `buildSchema` for the `Apollo.server`. How can I generate the schema file just from the types? (and not a resolver?) Otherwise my `client` has to depend on the `server` just for the graphql file, but I want all the graphql schema information in the `schema` package.

Alternatively, is there a way I can add the resolver definitions to the `schema` package, but leave the actual implementation to the `server` package?

**Reply by MichalLytek** [reply]
`buildSchema` with `emitSchemaFile` generates SDL. Then you can use any graphql tool that uses SDL - the same as for apollo, etc. No special TypeGraphQL treatment needed.
